ext/drbbdup/drbbdup.h

```diff
@@ -0,0 +1,334 @@
+ *
+ * @param instr the instr to be instrumented.
+ * @param where the location where to insert instrumentation.
+ * @param encoding the encoding of the current case considered.
+ * @param pre_analysis_data preliminary analysis data.
+ * @param analysis_data analysis data for the particular case.
+ */
+typedef void (*drbbdup_instrument_instr_t)(void *drcontext, instrlist_t *bb,
+                                           instr_t *instr, instr_t *where,
+                                           uintptr_t encoding, void *user_data,
+                                           void *pre_analysis_data, void *analysis_data);
+
+/***************************************************************************
+ * INIT
+ */
+
+/**
+ * Specifies the options when initialising drbbdup. \p set_up_bb_dups, \p insert_encode
+ * and \p instrument_instr cannot be NULL, while \p dup_limit must be greater than zero.
+ */
+typedef struct {
+    /** Set this to the size of this structure. */
+    size_t struct_size;
+    /** A user-defined call-back function that sets up how to duplicate a basic block.
+     * Cannot be NULL.
+     */
+    drbbdup_set_up_bb_dups_t set_up_bb_dups;
+    /**
+     * A user-defined call-back function that inserts code to encode the runtime case.
+     * The resulting encoding is used by the dispatcher to direct control to the
+     * appropriate basic block. Cannot be NULL.
+     */
+    drbbdup_insert_encode_t insert_encode;
+    /**
+     * A user-defined call-back function that conducts a preliminary analysis on a basic
+     * block.
+     */
+    drbbdup_pre_analyse_bb_t pre_analyse_bb;
+    /**
+     * A user-defined call-back function that destroys preliminary analysis data.
+     */
+    drbbdup_destroy_pre_analysis_t destroy_pre_analysis;
+    /**
+     * A user-defined call-back function that analyses a basic block for a particular
+     * case.
+     */
+    drbbdup_analyse_bb_t analyse_bb;
+    /**
+     * A user-defined call-back function that destroys analysis data for a particular
+     * case.
+     */
+    drbbdup_destroy_analysis_t destroy_analysis;
+    /**
+     * A user-defined call-back function that instruments an instruction with respect to a
+     * particular case.
+     */
+    drbbdup_instrument_instr_t instrument_instr;
+    /* User-data made available to user-defined call-back functions that drbbdup invokes
+     * to manage basic block duplication.
+     */
+    void *user_data;
+    /* The maximum number of cases, excluding the default case, that can be associated
+     * with a basic block. Once the limit is reached and an unhandled case is encountered,
+     * control is directed to the default case.
+     */
+    ushort dup_limit;
+} drbbdup_options_t;
+
+/**
+ * Priorities of drmgr instrumentation passes used by drbbdup. Users
+ * can perform app2app manipulations prior duplication
+ * by ordering such changes before DRMGR_PRIORITY_DRBBDUP.
```

Comment:
Use # to make a link

Reply:
Link to what please?

Reply:
NVM, I take it DRMGR_PRIORITY_DRBBDUP

Reply:
Yes, please see the doxygen notes in https://github.com/DynamoRIO/dynamorio/wiki/Code-Style which talk about `#` and `()`, and if you look at the produced html you can verify whether there's a link or not.